### PR TITLE
Add a user directory

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/handlers/user_directory.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        from not matching `public_value` to doing so.
+
+        For example, check if `history_visibility` (`key_name`) changed from
+        `shared` to `world_readable` (`public_value`).
+
+        Returns:
+            None if the field in the events either both match `public_value` o
+            neither do, i.e. there has been no change.
+            True if it didnt match `public_value` but now does
+            Falsse if it did match `public_value` but now doesn't
```
Comment (NegativeMjark) [inline synapse/handlers/user_directory.py]: > Falsse

File: synapse/handlers/user_directory.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    @defer.inlineCallbacks
+    def _get_key_change(self, prev_event_id, event_id, key_name, public_value):
+        """Given two events check if the `key_name` field in content changed
+        from not matching `public_value` to doing so.
+
+        For example, check if `history_visibility` (`key_name`) changed from
+        `shared` to `world_readable` (`public_value`).
+
+        Returns:
+            None if the field in the events either both match `public_value` o
```
Comment (NegativeMjark) [inline synapse/handlers/user_directory.py]: o

File: synapse/handlers/user_directory.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+            prev_event = yield self.store.get_event(prev_event_id, allow_none=True)
+
+        if event_id:
+            event = yield self.store.get_event(event_id, allow_none=True)
+
+        if not event and not prev_event:
+            defer.returnValue(None)
+
+        prev_hist_vis = None
+        hist_vis = None
```
Comment (NegativeMjark) [inline synapse/handlers/user_directory.py]: > hist_vis

File: synapse/handlers/user_directory.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+                    continue
+                elif not change and not is_public:
+                    continue
+
+                users_with_profile = yield self.state.get_current_user_in_room(room_id)
+                for user_id, profile in users_with_profile.iteritems():
+                    if change:
+                        yield self._handle_new_user(room_id, user_id, profile)
+                    else:
+                        yield self._handle_remove_user(room_id, user_id)
```
Comment (NegativeMjark) [inline synapse/handlers/user_directory.py]: It looks like lines 204-221 are duplicates of 172-197. Maybe lift them into a function?

File: synapse/storage/schema/delta/42/user_dir.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+INSERT INTO user_directory_stream_pos (stream_id) VALUES (null);
+
+CREATE TABLE user_directory (
+    user_id TEXT NOT NULL,
+    room_id TEXT NOT NULL,  -- A room_id that we know is public
+    display_name TEXT,
+    avatar_url TEXT
+);
+
+CREATE INDEX user_directory_room_idx ON user_directory(room_id);
```
Comment (NegativeMjark) [inline synapse/storage/schema/delta/42/user_dir.py]: Maybe add a UNIQUE index.

File: synapse/storage/user_directory.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        def get_current_state_deltas_txn(txn):
+            # First we calculate the max stream id that will give us less than
+            # N results
+            sql = """
+                SELECT stream_id, count(*)
+                FROM current_state_delta_stream
+                WHERE stream_id > ?
+                GROUP BY stream_id
+                ORDER BY stream_id ASC
+                LIMIT 100
```
Comment (NegativeMjark) [inline synapse/storage/user_directory.py]: Why 100?

File: synapse/storage/schema/delta/42/user_dir.py (hunk 5 above)
Comment (erikjohnston) [inline synapse/storage/schema/delta/42/user_dir.py]: Probably not on room_id, no.

But I've added one to user_id.

File: synapse/storage/user_directory.py (hunk 6 above)
Comment (erikjohnston) [inline synapse/storage/user_directory.py]: Done

File: synapse/handlers/user_directory.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    def _get_key_change(self, prev_event_id, event_id, key_name, public_value):
+        """Given two events check if the `key_name` field in content changed
+        from not matching `public_value` to doing so.
+
+        For example, check if `history_visibility` (`key_name`) changed from
+        `shared` to `world_readable` (`public_value`).
+
+        Returns:
+            None if the field in the events either both match `public_value`
+            neither do, i.e. there has been no change.
```
Comment (NegativeMjark) [inline synapse/handlers/user_directory.py]: Should this be "or if neither do" ?

File: synapse/handlers/user_directory.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+        if event:
+            value = event.content.get(key_name, None)
+
+        if value == public_value and prev_value != public_value:
+            defer.returnValue(True)
+        elif value != public_value and prev_value == public_value:
+            defer.returnValue(False)
+        else:
+            defer.returnValue(None)
```
Comment (NegativeMjark) [inline synapse/handlers/user_directory.py]: You could write this as:

```python
old_value_is_public = prev_value == public_value
new_value_is_public = value == public_value
if old_value_is_public == new_value_is_public:
   defer.returnValue(None)
else:
   defer.returnValue(new_value_is_public)
```

I'm not sure if it is clearer or not.

File: synapse/handlers/user_directory.py [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+
+        prev_value = None
+        value = None
+
+        if prev_event:
+            prev_value = prev_event.content.get(key_name, None)
+
+        if event:
+            value = event.content.get(key_name, None)
```
Comment (NegativeMjark) [inline synapse/handlers/user_directory.py]: The `None` is redundant.

File: synapse/handlers/user_directory.py (hunk 8 above)
Comment (erikjohnston) [inline synapse/handlers/user_directory.py]: I don't find it any more readable tbh

**Comment by mercenaryship** [comment]
Close down 711 for selling pirated cigarettes. And raid the stores selling
opium laced cigarettes. Subsidize real tobacco growers which deliver
tobacco, real tobacco.

ALL RIGHTS RESERVED

**Pull request merged**